fix: Cancel PipelineRuns on MR change in Gitlab

aThorp96 opened the pull request:
# Changes <!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

- Remove seemingly redundant label selector when cancelling PipelineRuns for a Pull Request Change which was incompatible with GitLab Merge Requests. 
- Consolidated git repo initialization in tests

# Submitter Checklist

- [ ] 📝 Ensure your commit message is clear and informative. Refer to the How to write a git commit message guide. Include the commit message in the PR body rather than linking to an external site (e.g., Jira ticket).

- [x] ♽ Run make test lint before submitting a PR to avoid unnecessary CI processing. Consider installing [pre-commit](https://pre-commit.com/) and running pre-commit install in the repository root for an efficient workflow.

- [x] ✨ We use linters to maintain clean and consistent code. Run make lint before submitting a PR. Some linters offer a --fix mode, executable with make fix-linters (ensure [markdownlint](https://github.com/DavidAnson/markdownlint) and [golangci-lint](https://github.com/golangci/golangci-lint) are installed).

- [ ] 📖 Document any user-facing features or changes in behavior.

- [x] 🧪 While 100% coverage isn't required, we encourage unit tests for code changes where possible.

- [x] 🎁 If feasible, add an end-to-end test. See [README](https://github.com/openshift-pipelines/pipelines-as-code/blob/main/test/README.md) for details.

- [x] 🔎 Address any CI test flakiness before merging, or provide a valid reason to bypass it (e.g., token rate limitations).

- If adding a provider feature, fill in the following details:

| Git Provider     | Supported |
|------------------|-----------|
| GitHub App       | ✅️        |
| GitHub Webhook   |  ✅️        |
| Gitea            | ❌️        |
| GitLab           |  ✅️        |
| Bitbucket Cloud  | ❌️        |
| Bitbucket Server | ❌️        |

  (update the documentation accordingly)

aThorp96 committed:
fix: Cancel PipelineRuns on MR change in Gitlab

Signed-off-by: Andrew Thorp <athorp@redhat.com>

chmouel commented:
/retest

aThorp96 commented on test/pkg/scm/scm.go:
```diff
@@ -48,7 +97,12 @@ func gitPushPullRetry(t *testing.T, opts *Opts, path string) {
-			return
+
+			// get sha
+			sha, err := git.RunGit(path, "rev-parse", "HEAD")
```
Moving the `rev-parse` to here as this function may change the sha if it performs a rebase

chmouel commented:
the change is good, but i think there is an opportunity to clean this code
I think cancelAllInProgressBelongingToPullRequest should be used instead since it already does that
and move the cancel logic to a new method for push, i wrote the patch here and let me know what do you think:


```diff
diff --git a/pkg/pipelineascode/cancel_pipelineruns.go b/pkg/pipelineascode/cancel_pipelineruns.go
index 064fc4b0..f335e909 100644
--- a/pkg/pipelineascode/cancel_pipelineruns.go
+++ b/pkg/pipelineascode/cancel_pipelineruns.go
@@ -55,13 +55,7 @@ func (p *PacRun) cancelAllInProgressBelongingToPullRequest(ctx context.Context,
 	return nil
 }
 
-// cancelInProgressMatchingPR cancels all PipelineRuns associated with a given repository and pull request,
-// except for the one that triggered the cancellation. It first checks if the cancellation is in progress
-// and if the repository has a concurrency limit. If a concurrency limit is set, it returns an error as
-// cancellation is not supported with concurrency limits. It then retrieves the original pull request name
-// from the annotations and lists all PipelineRuns with matching labels. For each PipelineRun that is not
-// already done, cancelled, or gracefully stopped, it patches the PipelineRun to cancel it.
-func (p *PacRun) cancelInProgressMatchingPR(ctx context.Context, matchPR *tektonv1.PipelineRun, repo *v1alpha1.Repository) error {
+func (p *PacRun) cancelAllInProgressBelongingToPush(ctx context.Context, matchPR *tektonv1.PipelineRun, repo *v1alpha1.Repository) error {
 	if matchPR == nil {
 		return nil
 	}
@@ -82,9 +76,6 @@ func (p *PacRun) cancelInProgressMatchingPR(ctx context.Context, matchPR *tekton
 		keys.URLRepository:  formatting.CleanValueKubernetes(p.event.Repository),
 		keys.OriginalPRName: prName,
 	}
-	if p.event.TriggerTarget == triggertype.PullRequest {
-		labelMap[keys.PullRequest] = strconv.Itoa(p.event.PullRequestNumber)
-	}
 	labelSelector := getLabelSelector(labelMap)
 	p.run.Clients.Log.Infof("cancel-in-progress: selecting pipelineRuns to cancel with labels: %v", labelSelector)
 	prs, err := p.run.Clients.Tekton.TektonV1().PipelineRuns(matchPR.GetNamespace()).List(ctx, metav1.ListOptions{
@@ -112,6 +103,22 @@ func (p *PacRun) cancelInProgressMatchingPR(ctx context.Context, matchPR *tekton
 	return nil
 }
 
+// cancelInProgressMatchingPR cancels all PipelineRuns associated with a given repository and pull request,
+// except for the one that triggered the cancellation. It first checks if the cancellation is in progress
+// and if the repository has a concurrency limit. If a concurrency limit is set, it returns an error as
+// cancellation is not supported with concurrency limits. It then retrieves the original pull request name
+// from the annotations and lists all PipelineRuns with matching labels. For each PipelineRun that is not
+// already done, cancelled, or gracefully stopped, it patches the PipelineRun to cancel it.
+func (p *PacRun) cancelInProgressMatchingPR(ctx context.Context, matchPR *tektonv1.PipelineRun, repo *v1alpha1.Repository) error {
+	if p.event.TriggerTarget == triggertype.PullRequest {
+		return p.cancelAllInProgressBelongingToPullRequest(ctx, repo)
+	} else if p.event.TriggerTarget == triggertype.Push {
+		return p.cancelAllInProgressBelongingToPush(ctx, matchPR, repo)
+	}
+
+	return fmt.Errorf("event %s is not supported for canceling in progress", p.event.TriggerTarget.String())
+}
+
 // cancelPipelineRunsOpsComment cancels all PipelineRuns associated with a given repository and pull request.
 // when the user issue a cancel comment.
 func (p *PacRun) cancelPipelineRunsOpsComment(ctx context.Context, repo *v1alpha1.Repository) error {
 ```

aThorp96 commented:
@chmouel I think there is opportunity for refactoring here, but I'm not sure if splitting on the `push` and `pull-request` is the right move here. `cancelAllInProgressBelongingToPullRequest` doesn't work as-is because since it cancels all PipelineRuns associated with the PR it will also cancel the PipelineRun which triggered the event. The current logic is split on whether or not the cancel-function should skip an existing PipelineRun. I will think on how we may be able to better consolidate and delineate the existing logic.

aThorp96 committed:
chore: experimental refactor

Experimental refactor of cancel-in-progress label selection. Kept as a
separate commit for now to more easily scrap, if abandoned

aThorp96 commented:
@chmouel I just pushed a [separate commit](https://github.com/openshift-pipelines/pipelines-as-code/pull/1954/commits/37d177c8a3ea3bae86d5c0bed91433496fda390e) which did some refactoring and consolidation. LMK what you think? If it's acceptable I'll squash it

chmouel commented:
is the e2e failure related?

aThorp96 commented:
/retest

aThorp96 commented:
@chmouel I think it is flaky. I can't reproduce it locally

aThorp96 commented:
d'oh, I guess `/retest` doesn't retry the e2e tests and I don't have the permissions to retry via the Github UI. Good thing I needed to rebase off main I guess 😅

chmouel commented:
I have sent you an invite to add you to the pac group

aThorp96 commented:
@chmouel is this good to rebase+squash and then merge?

chmouel commented:
/merge rebase

chmouel commented:
/merge rebase

chmouel commented:
/merge
